Add bit-precise overloads for builtin operators

AaronBallman opened the pull request:
We previously were not adding them to the candidate set and so use of a bit-precise integer as a class member could lead to ambiguous overload sets.

Fixes https://github.com/llvm/llvm-project/issues/82998

AaronBallman committed:
Add bit-precise overloads for builtin operators

We previously were not adding them to the candidate set and so use of a
bit-precise integer as a class member could lead to ambiguous overload
sets.

Fixes #82998

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-bitint-builtin-operators

erichkeane commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -8955,6 +8962,22 @@ class BuiltinOperatorOverloadBuilder {
+      for (QualType BitTy : Candidate.bitint_types())
+        BitIntCandidates.insert(CanQualType::CreateUnsafe(BitTy));
+    });
+    llvm::copy(BitIntCandidates, std::back_inserter(ArithmeticTypes));
```
I'm concerned about the assert on `9001`.  It seems to me that this changes us from being a reasonably 'fixed' list to having a perhaps-unlimited list.  

I think that the intent of that assert is nice (ensuring we can small-vector-opt this list), but this section here makes that no longer guaranteed, right? 

The `ArithmeticTypesCap` is 24, and we do a `push_back` of 23 other types (by my count!).  So adding TWO bit-int types + having all the rest of the types enabled causes that assert to fire.

AaronBallman replied:
Done

AaronBallman committed:
Small optimization to use move instead of copy

AaronBallman committed:
Improve comment based on review feedback; NFC

AaronBallman committed:
Account for bit-precise types in an assert

We're subtracting out the bit-precise types when determining whether
we're under the cap because we don't know for sure how many bit-precise
types will wind up in the overload set. This assertion exists to let us
know when we've hit a point where we need to allocate rather than use
the small vector inline storage, so to avoid needing those allocations,
I've bumped the cap up by two elements on the assumption that builtin
overload set will often not have more than two candidate types.

AaronBallman commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -8955,6 +8962,22 @@ class BuiltinOperatorOverloadBuilder {
+      for (QualType BitTy : Candidate.bitint_types())
+        BitIntCandidates.insert(CanQualType::CreateUnsafe(BitTy));
+    });
+    llvm::copy(BitIntCandidates, std::back_inserter(ArithmeticTypes));
```
Done

AaronBallman commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -8955,6 +8962,22 @@ class BuiltinOperatorOverloadBuilder {
+    /// _BitInt overloads are a bit special. We don't want to add candidates
+    /// for the entire family of _BitInt types, so instead we only add
+    /// candidates for the unique, unqualified _BitInt types present in the
+    /// candidate type set. The candidate set already handled ensuring the type
```
Done

erichkeane reviewed (approved):
LGTM if Corentin is happy with the comment.

cor3ntin reviewed (approved):
LGTM, thanks!
